pkg/sparrow/run.go

```diff
@@ -290,13 +298,40 @@ func fanInResults(checkChan chan checks.Result, cResult chan checks.ResultDTO, n
 // shutdown shuts down the sparrow and all managed components gracefully.
 // It returns an error if one is present in the context or if any of the
 // components fail to shut down.
-func (s *Sparrow) shutdown(ctx context.Context) error {
+func (s *Sparrow) shutdown(ctx context.Context) {
 	errC := ctx.Err()
+	log := logger.FromContext(ctx)
 	ctx, cancel := context.WithTimeout(context.Background(), shutdownTimeout)
 	defer cancel()
-	errS := s.targets.Shutdown(ctx)
-	if errS != nil {
-		return fmt.Errorf("failed to shutdown sparrow: %w", errors.Join(errC, errS))
+
+	var errS error
+	if s.cfg.HasTargetManager() {
+		errS = s.targets.Shutdown(ctx)
+	}
+	errA := s.shutdownAPI(ctx)
+	if errS != nil || errA != nil {
+		log.Error("Failed to shutdown gracefully", "contextError", errC, "apiError", errA, "targetError", errS)
+	}
+
+	s.cDone <- struct{}{}
+}
+
+// handleErrors handles errors from the sparrow components
+// If a non-recoverable error happens, the sparrow will shut down
+// the various components and return the error.
+func (s *Sparrow) handleErrors(ctx context.Context) {
+	log := logger.FromContext(ctx)
+	for {
```

Comment:
Remove the for loop, otherwise multiple errors in cErr could cause shutdown to be called multiple times

Reply:
I just used the native go way of doing this, by adding a `sync.Once` in the shutdown function itself.